lib/course_work_courses.rb

```diff
@@ -110,14 +88,8 @@ def find_by_class_id_and_section_and_sunet(class_id, section, sunet)
   # TODO: We have tests that are highly sensitive to the order of the courses; this unfortunate
   # logic preserves the bottom-most course from the xml. it's unclear whether this is incidental
   # or a feature.
-  # Breaking out into two approaches.  JSON is picked only if the Setting is present
   def all_courses
-    # if Settings.use_course_json == true
-    if Settings.use_course_json
-      @all_courses ||= process_all_courses(self.json_files).to_a.reverse.uniq(&:key).reverse.to_a
-    else
-      @all_courses ||= process_all_courses_xml(self.raw_xml).to_a.reverse.uniq(&:key).reverse.to_a
-    end
+    @all_courses ||= process_all_courses(self.json_files).to_a.reverse.uniq(&:key).reverse.to_a
```

Comment:
Why do we need a double reverse here? Why does it need to be cast to `.to_a` a second time?

Reply:
This is a copy of the original handling. I am not sure why it was included beyond the original comment above the method "We have tests that are highly sensitive to the order of the courses; this unfortunate logic preserves the bottom-most course from the xml. it's unclear whether this is incidental or a feature." I can look into this further.

Reply:
If it doesn't have a test, then it's not a feature 😄

Reply:
There is a specific test that fails when I try to change the code here to simply use .uniq(&:key) instead: https://github.com/sul-dlss/course_reserves/blob/master/spec/lib/course_work_courses_spec.rb#L165 .  The reason is that, in this test, the original list of courses has two sections that are identical with respect to their CIDs and their instructor sunet id list.  They only differ with respect to their section number.  The test expects that, when it looks for a course with that CID, there will only be one section returned with the lower section number (i.e. 1).  Based on the fixture file, the first reverse, I think, is intended to put the sections in ascending order, with uniq using the Course class's key method (which is based on CID and instructor SUNET IDs), returning the very first section with the same course id and instructors (and removing the rest), and then the next reverse puts the list back in the original order.  

When I tried a sort by section id followed by using uniq based on the key method, it changes the original order of the courses, which make other tests fail.
I could write a separate method which does what this line of code is trying to do (preserve the first section among courses that have the same CID and same set of instructors, while removing the rest), while preserving the original order of the courses.

Reply:
I am not sure why the section number matters to these tests in this way, because the app only displays the course information without section information included, but I do see section id is referenced in other code, and I don't want to change this part without understanding the consequences.

Reply:
I've added another method which does have the tests pass.  If there's a better/more concise way to do that, please feel free to suggest.